### PR TITLE
handle AsyncAfterWriteAction exception before listener is registered

### PR DESCRIPTION
**Pull request by vladimirdolzhenko**
AsyncAfterWriteAction error handling improvement, part 1
Relates to #31716

- any exceptions in `indexShard.sync(location, syncListener)` could be swallowed - as it could be done asynchronously and synchronously - at the moment `listener` could not be registered.

### CODE REVIEW

**Review by bleskes — CHANGES_REQUESTED** [review]
Thx Vlad. I left some minor comments. Can you also change the PR title to better describe what the problem is? 

I'm a bit uncomfortable merging this before we have a complete solution to #31716 as it may erroneously fail indexing requests (it's likely that the async write actions finish before the replication operation is completed - which is a situation we were not reporting the error so far). It's good we're doing this in small steps, but how about merging this into a feature branch where we can iterate towards a complete solution? we can then merge it into master in one go and it will also be easier to back port (I think this can go into 6.x, if we do the right thing)

File: server/src/test/java/org/elasticsearch/action/support/replication/TransportWriteActionTests.java [hunk 1]
```diff
@@ -158,6 +160,33 @@ public void testReplicaNoRefreshCall() throws Exception {
...
+        TestAction testAction = new TestAction();
+        TransportWriteAction.WritePrimaryResult<TestRequest, TestResponse> result =
+            testAction.shardOperationOnPrimary(request, indexShard);
+
+        final Consumer<Exception> syncListener = syncListenerRef.get();
+        assertThat(syncListener, notNullValue());
+
+        syncListener.accept(new IOException("xxx"));
+
+        CapturingActionListener<TestResponse> listener = new CapturingActionListener<>();
```
[the remark this answers is not in the document]
Comment (vladimirdolzhenko) [reply server/src/test/java/org/elasticsearch/action/support/replication/TransportWriteActionTests.java]: agree

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 2]
```diff
@@ -2328,7 +2328,7 @@ protected void write(List<Tuple<Translog.Location, Consumer<Exception>>> candida
      * NOTE: if the syncListener throws an exception when it's processed the exception will only be logged. Users should make sure that the
      * listener handles all exception cases internally.
      */
-    public final void sync(Translog.Location location, Consumer<Exception> syncListener) {
+    public void sync(Translog.Location location, Consumer<Exception> syncListener) {
```
[the remark this answers is not in the document]
Comment (vladimirdolzhenko) [reply server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: yes, it's required for mocking

File: server/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java [hunk 3]
```diff
@@ -186,11 +189,14 @@ public synchronized void onSuccess(boolean forcedRefresh) {
...
      */
-    protected static class WriteReplicaResult<ReplicaRequest extends ReplicatedWriteRequest<ReplicaRequest>>
+    public static class WriteReplicaResult<ReplicaRequest extends ReplicatedWriteRequest<ReplicaRequest>>
             extends ReplicaResult implements RespondingWriteResult {
         public final Location location;
-        boolean finishedAsyncActions;
-        private ActionListener<TransportResponse.Empty> listener;
+        // as AsyncAfterWriteAction could execute on another thread
+        private volatile boolean finishedAsyncActions;
+        private volatile Exception failure;
```
[the remark this answers is not in the document]
Comment (vladimirdolzhenko) [reply server/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java]: there is `public final Exception finalFailure;` in `PrimaryResult` (the same in `ReplicaResult`) - shall we consider to merge them ? they do have slightly different (but close) semantic

**Comment by vladimirdolzhenko** [comment]
thanks @bleskes for comments - I did required changes.

In fact https://github.com/elastic/elasticsearch/pull/31857 (as a complete solution for #31716) is based on this PR.

File: server/src/test/java/org/elasticsearch/index/replication/AsyncAfterWriteActionTests.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+            if ((routing.primary() && failOnPrimary)
+                || (routing.primary() == false && failOnPrimary == false)) {
+                exceptionCount.incrementAndGet();
+                return ioException;
+            }
+            return null;
+        };
+        try (ReplicationGroup shards = createGroup(0, exceptionFunction)) {
+            shards.startAll();
+            if (randomBoolean()) {
```
Comment (bleskes) [inline server/src/test/java/org/elasticsearch/index/replication/AsyncAfterWriteActionTests.java]: this is not needed, right?

**Review by bleskes — APPROVED** [review]
I left some a minor comment on the test that isn't required for another review. Can you please merge this into a *feature* branch and proceed with prs to that?

**Comment by vladimirdolzhenko** [comment]
merged into [feature/31755](https://github.com/elastic/elasticsearch/tree/feature/31755)

**Pull request closed without merging**